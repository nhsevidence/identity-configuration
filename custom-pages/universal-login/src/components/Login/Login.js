import React, { Component } from "react";
import { Alert } from '@nice-digital/nds-alert';
import pathOr from 'ramda/src/pathOr';
import { Input } from '@nice-digital/nds-forms';
import { FormGroup } from '@nice-digital/nds-form-group';
import qs from 'qs';
import { Link } from "react-router-dom";
import Nav from "../Nav/Nav";
import { isDomainInUsername, showNav} from '../../helpers';
import AuthApi from '../../services/AuthApi';
import { auth as authOpts } from '../../services/constants';
import './Login.scss';

class Login extends Component {
  constructor(props) {
    super(props);
    this.auth = new AuthApi();
    this.state = {
      username: null,
      password: null,
      error: null,
      loading: false,
      isAD: false,
      connection: authOpts.connection,
      showGoogleLogin: false,
      activationEmailSent: false
    };
    this.querystring = qs.parse(document.location.search, {
      ignoreQueryPrefix: true
    });
    this.continue = true
  }

  componentDidMount() {
    this.querystring = qs.parse(document.location.search, {
      ignoreQueryPrefix: true
    });
    this.auth.fetchClientSettings().then(() => {
      this.googleConnection = pathOr(
        null,
        ['strategies', 'google-oauth2', 'connectionName'],
        window.Auth0
      );
      this.ADConnection = pathOr(
        null,
        ['strategies', 'waad', 'connectionName'],
        window.Auth0
      );
      this.setState(
        { showGoogleLogin: !!this.googleConnection },
        this.showAuth0RulesError
      )
    })
  }

  showAuth0RulesError = () => {
    this.setState({ error: this.querystring.myerror })
  };

  resendVerificationEmail = (e) => {
    if (e) e.preventDefault()
    const callback = err =>
      this.setState({ activationEmailSent: !err, error: err })
    try {
      this.auth.resendVerificationEmail(this.querystring.userid, callback)
    } catch (err) {
      console.log(JSON.stringify(err))
    }
  };

  login = (e, isGoogle) => {
    if (e) e.preventDefault()
    const requestErrorCallback = err =>
      this.setState(
        {
          error: err.description || err.error_description,
          loading: false
        },
        console.log(JSON.stringify(err))
      );
    try {
      this.setState({ loading: true }, () => {
        const { username, password, connection } = this.state
        const loginConnection = isGoogle ? this.googleConnection : connection
        const isResumingAuthState =
          this.querystring.myerrorcode &&
          this.querystring.myerrorcode === 'user_not_verified'
            ? this.continue && this.querystring.state
            : null
        this.auth.login(
          loginConnection,
          username,
          password,
          requestErrorCallback,
          isResumingAuthState
        )
      })
    } catch (err) {
      console.log(JSON.stringify(err))
      this.setState({ loading: false, error: 'Something has gone wrong.' })
    }
  };

  handleChange = ({ target: { name, value } }) => {
    let isAD = null;
    if (name === 'username') {
      isAD = isDomainInUsername(value);
      if (this.querystring.myerrorcode && this.querystring.email !== value) {
        this.continue = false
      }
    }
    this.setState({
      [name]: value,
      // error: null,
      isAD,
      connection: isAD ? this.ADConnection : authOpts.connection
    })
  };

  render() {
    showNav();
    const {
      error,
      loading,
      isAD,
      showGoogleLogin,
      activationEmailSent
    } = this.state
    const { myerrorcode } = this.querystring;

    return (
      <div>
        <Nav/>
        <form className="">
          <FormGroup legend="Personal information">
            {error && (
              <Alert type="error">
                {error}{' '}
                {myerrorcode === 'user_not_verified' ? (
                  <button href="#" onClick={this.resendVerificationEmail}>
                    Resend activation email
                  </button>
                ) : null}
              </Alert>
            )}
            {activationEmailSent && (
              <Alert type="success">An activation email has been sent!</Alert>
            )}
            <Input
              data-qa-sel="login-email"
              label="Email"
              id="username"
              name="username"
              type="email"
              placeholder="eg: your.name@example.com..."
              onChange={this.handleChange}
              autoComplete="username"
            />
            {!isAD && (
              <Input
                data-qa-sel="login-password"
                name="password"
                type="password"
                label="Password"
                onChange={this.handleChange}
                autoComplete="current-password"
              />
            )}
          </FormGroup>
          {!loading ? (
            <div>
              <button
                data-qa-sel="login-button"
                className="btn btn--cta"
                onClick={e => this.login(e, false)}
                // disabled={!username}
              >
                Sign in
              </button>
              {showGoogleLogin && (
                <button
                  data-qa-sel="login-button-social"
                  className="iconBtn social"
                  style={{ float: 'right' }}
                  onClick={e => this.login(e, true)}
                >
                  <span className="buttonLabel">Or </span>
                  <img
                    className="iconBtn-icon"
                    alt="Sign in with google"
                    src={ process.env.PUBLIC_URL + "/images/btn_google_signin_light_normal_web.png"}
                  />
                </button>
              )}
            </div>
          ) : (
            'Loading...'
          )}
        </form>
        <Link className="forgotPasswordLink" data-qa-sel="forgotPassword-link" to="/forgotPassword">
          Forgot password?
        </Link>
      </div>
    )
  }
}

export default Login