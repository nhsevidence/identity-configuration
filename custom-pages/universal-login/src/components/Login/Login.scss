.mainpanel {
  display: flex;
  flex-direction: column;
  // width: 30%;
  margin: auto;
  min-width: var(--col-width-mobile);
}
.iconBtn{
  border:none;
  background: none;
}
.iconBtn-icon:hover{
  opacity: 0.8;
}
.buttonLabel{
  font-weight: 700;
  padding-right: 6px;
}
