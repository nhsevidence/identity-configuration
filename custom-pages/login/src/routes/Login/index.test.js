import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-14'

import Login from './'

Enzyme.configure({ adapter: new Adapter() })

describe('Login components', () => {
  jest.mock('../../services/AuthApi', () => jest.fn(() => {}))

  it('should render <Login /> correctly', () => {
    expect(shallow(<Login.component />)).toMatchSnapshot()
  })
})
