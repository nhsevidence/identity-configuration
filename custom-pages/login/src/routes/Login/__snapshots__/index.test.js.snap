// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login components should render <Login /> correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Fieldset
        legend="Personal Information"
      >
        <Input
          id="username"
          label="Username"
          name="username"
          onChange={[Function]}
          placeholder="eg: your.name@example.com..."
          type="email"
        />
        <Input
          label="Password"
          name="password"
          onChange={[Function]}
          type="password"
        />
        <button
          className="btn btn--cta"
          disabled={true}
          onClick={[Function]}
        >
          Sign in
        </button>
      </Fieldset>,
      "className": "",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          null,
          <Input
            id="username"
            label="Username"
            name="username"
            onChange={[Function]}
            placeholder="eg: your.name@example.com..."
            type="email"
          />,
          <Input
            label="Password"
            name="password"
            onChange={[Function]}
            type="password"
          />,
          <button
            className="btn btn--cta"
            disabled={true}
            onClick={[Function]}
          >
            Sign in
          </button>,
        ],
        "legend": "Personal Information",
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "username",
            "label": "Username",
            "name": "username",
            "onChange": [Function],
            "placeholder": "eg: your.name@example.com...",
            "type": "email",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "label": "Password",
            "name": "password",
            "onChange": [Function],
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign in",
            "className": "btn btn--cta",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Sign in",
          "type": "button",
        },
      ],
      "type": [Function],
    },
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Fieldset
          legend="Personal Information"
        >
          <Input
            id="username"
            label="Username"
            name="username"
            onChange={[Function]}
            placeholder="eg: your.name@example.com..."
            type="email"
          />
          <Input
            label="Password"
            name="password"
            onChange={[Function]}
            type="password"
          />
          <button
            className="btn btn--cta"
            disabled={true}
            onClick={[Function]}
          >
            Sign in
          </button>
        </Fieldset>,
        "className": "",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            null,
            <Input
              id="username"
              label="Username"
              name="username"
              onChange={[Function]}
              placeholder="eg: your.name@example.com..."
              type="email"
            />,
            <Input
              label="Password"
              name="password"
              onChange={[Function]}
              type="password"
            />,
            <button
              className="btn btn--cta"
              disabled={true}
              onClick={[Function]}
            >
              Sign in
            </button>,
          ],
          "legend": "Personal Information",
        },
        "ref": null,
        "rendered": Array [
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "id": "username",
              "label": "Username",
              "name": "username",
              "onChange": [Function],
              "placeholder": "eg: your.name@example.com...",
              "type": "email",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "label": "Password",
              "name": "password",
              "onChange": [Function],
              "type": "password",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign in",
              "className": "btn btn--cta",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Sign in",
            "type": "button",
          },
        ],
        "type": [Function],
      },
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFourteenAdapter {
      "options": Object {
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
          "getChildContext": Object {
            "calledByRenderer": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
